OrbitCore/LinuxEventTracer.cpp

```diff
@@ -99,150 +109,229 @@ void LinuxEventTracer::Run(
+                      header);
+
+              std::vector<unwindstack::FrameData> frames = unwinder.Unwind(
+                  sample.Registers(), sample.StackDump(), sample.StackSize());
+              if (!frames.empty()) {
+                HandleCallstack(sample.TID(), sample.Timestamp(), frames);
+              }
+
+              sample.SetFunction(
+                  uretprobe_fds_to_function.at(ring_buffer.FileDescriptor()));
+              uprobe_event_processor.Push(
+                  std::make_unique<LinuxUretprobeEventWithStack>(
+                      std::move(sample)));
 
-            std::vector<unwindstack::FrameData> frames = unwinder.Unwind(
-                sample.Registers(), sample.StackDump(), sample.StackSize());
-            if (!frames.empty()) {
-              HandleCallstack(sample.TID(), sample.Timestamp(), frames);
+            } else {
+              auto sample =
+                  ring_buffer.ConsumeRecord<LinuxStackSampleEvent>(header);
+
+              std::vector<unwindstack::FrameData> frames = unwinder.Unwind(
+                  sample.Registers(), sample.StackDump(), sample.StackSize());
+              if (!frames.empty()) {
+                HandleCallstack(sample.TID(), sample.Timestamp(), frames);
+              }
             }
           } break;
 
           case PERF_RECORD_LOST: {
             auto lost = ring_buffer.ConsumeRecord<LinuxPerfLostEvent>(header);
-            PRINT("Lost %u Events\n", lost.Lost());
+            PRINT("Lost %u events\n", lost.Lost());
           } break;
 
-          default:
-            PRINT("Unexpected Perf Sample Type: %u\n", header.type);
+          default: {
+            PRINT("Unexpected perf_event_header::type: %u\n", header.type);
             ring_buffer.SkipRecord(header);
-            break;
+          } break;
         }
       }
     }
+
+    if (new_thread >= 0) {
+      int32_t fd =
+          LinuxPerfUtils::stack_sample_event_open(new_thread, period_ns_);
+      fds.push_back(fd);
+      ring_buffers.emplace_back(fd);
+      LinuxPerfUtils::start_capturing(fd);
+    }
+
+    uprobe_event_processor.ProcessTillOffset();
   }
 
+  uprobe_event_processor.ProcessAll();
+
   // stop capturing
   for (int32_t fd : fds) {
     LinuxPerfUtils::stop_capturing(fd);
   }
+
   ring_buffers.clear();
   for (int32_t fd : fds) {
     close(fd);
   }
 }
 
-//-----------------------------------------------------------------------------
-void LinuxEventTracer::HandleCallstack(
-    ThreadID tid, uint64_t timestamp,
+bool LinuxEventTracerThread::ComputeFrequency() {
+  double period_ns_dbl = 1000 * 1000 * 1000 / frequency_;
+  if (period_ns_dbl > 0 &&
+      period_ns_dbl <
+          static_cast<double>(std::numeric_limits<uint64_t>::max())) {
+    period_ns_ = static_cast<uint64_t>(period_ns_dbl);
+    return true;
+  } else {
+    PRINT("Invalid frequency: %.3f\n", frequency_);
+    return false;
+  }
+}
+
+void LinuxEventTracerThread::LoadNumCpus() {
```

Comment:
I also think this name is misleading. I don't know what "loading num CPUs" means. Just "GetNumCpus" would be clearer, IMO.

Reply:
The reason this is not called Get... is that the result is not returned but assigned to a member. Like for the period computation this could be made entirely functional but I'm not sure of the benefit.